README
XML-OPDS

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc XML::OPDS

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=XML-OPDS

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/XML-OPDS

    CPAN Ratings
        http://cpanratings.perl.org/d/XML-OPDS

    Search CPAN
        http://search.cpan.org/dist/XML-OPDS/


LICENSE AND COPYRIGHT

Copyright (C) 2016 Marco Pessotto

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.



Changes
Revision history for XML-OPDS

0.04 2016-03-08

 - Fixed Moo syntax for older moos
 - Doc fixes

0.03 2016-03-06

 - Added support for logo and icon

0.02 2016-03-06

 - Added support for paging
 - Enforce RFC3339 format for updated fields

0.01 2016-03-05

 - Initial (and working) release.




Makefile.PL
use 5.008;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'XML::OPDS',
    AUTHOR           => q{Marco Pessotto <melmothx@gmail.com>},
    VERSION_FROM     => 'lib/XML/OPDS.pm',
    ABSTRACT_FROM    => 'lib/XML/OPDS.pm',
    LICENSE          => 'Perl_5',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.008,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
        'Test::Differences' => 0,
    },
    PREREQ_PM => {
                  'Moo' => 1,
                  'Type::Tiny' => 1,
                  'XML::Atom' => 0.41,
                  'DateTime' => 0,
                  'DateTime::Format::RFC3339' => 0,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'XML-OPDS-*' },
    META_MERGE => {
                   resources => {
                                 repository => 'https://github.com/melmothx/libxml-opds-perl',
                                 homepage => 'http://amusewiki.org',
                                 x_IRC => 'irc://chat.freenode.net/#amusewiki',
                                 x_WebIRC => 'https://webchat.freenode.net/?channels=#amusewiki',
                                },
                  },

);


lib/XML/OPDS/Navigation.pm
package XML::OPDS::Navigation;

use strict;
use warnings FATAL => 'all';
use Types::Standard qw/Str Enum Bool InstanceOf Object/;
use Moo;
use DateTime;
use XML::Atom::Link;

=head1 NAME

XML::OPDS::Navigation - Navigation elements for OPDS feeds

=head1 SETTERS/ACCESSORS

The all are read-write

=head2 prefix

If provided, every uri will have this string prepended, so you can
just pass URIs like '/path/to/file' and have them consistently turned
to 'http://myserver.org/path/to/file' if you set this to
'http://myserver.org'. See also L<XML::OPDS> C<prefix> method.

=head2 href

Required. The URI of the resource. If prefix is provided it is
prepended on output.

=head2 id

=head2 rel

Defaults to C<subsection>. Permitted values: C<self>, C<start>, C<up>,
C<subsection>, C<search>.

Additionally:

C<new> and C<popular> will expand to L<http://opds-spec.org/sort/new> and
L<http://opds-spec.org/sort/popular> as per spec paragraph 7.4.1.

Acquisition Feeds using the "http://opds-spec.org/sort/new" relation SHOULD be
ordered with the most recent items first. Acquisition Feeds using the "
http://opds-spec.org/sort/popular" relation SHOULD be ordered with the most
popular items first.

C<featured> will expand to L<http://opds-spec.org/featured> as per 7.4.2

C<recommended> will expand to L<http://opds-spec.org/recommended> as per 7.4.3

C<shelf> will expand to L<http://opds-spec.org/shelf> and
C<subscriptions> to L<http://opds-spec.org/subscriptions>.

C<crawlable> will expand to L<http://opds-spec.org/crawlable>.

This list is a work in progress and probably incomplete.

Facets are not supported yet (patches welcome). Client support for
facets is unclear. L<https://en.wikipedia.org/wiki/OPDS>.

=head2 title

=head2 acquistion

Boolean, default to false. Indicates that the C<href> is a leaf feed
with acquisition entries.

=head2 description

HTML allowed.

=head2  updated

A L<DateTime> object with the time of last update.

=head2 prefix



=cut

has id => (is => 'rw', isa => Str);

has rel => (is => 'rw',
            isa => Enum[qw/self start up subsection search/,
                        qw/first last previous next/, # RFC 5005
                        keys(%{ +{ __PACKAGE__->_rel_map } })],
            default => sub { 'subsection' });

has title => (is => 'rw', isa => Str);

has href => (is => 'rw', isa => Str, required => 1);

has acquisition => (is => 'rw', isa => Bool, default => sub { 0 });

has description => (is => 'rw', isa => Str);

has updated => (is => 'rw', isa => InstanceOf['DateTime'],
                default => sub { return DateTime->now });

has prefix => (is => 'rw', isa => Str, default => sub { '' });

has _dt_formatter => (is => 'ro', isa => Object, default => sub { DateTime::Format::RFC3339->new });

=head1 METHODS

The are mostly internals and used by L<XML::OPDS>

=head2 link_type

Depend if C<acquisition> is true of false.

=head2 as_link

The navigation as L<XML::Atom::Link> object.

=head2 identifier

Return the id or the URI.

=head2 relationship

[INTERNAL] Resolve the rel shortcuts.

=head2 as_entry

The navigation as L<XML::Atom::Entry> object.

=cut

sub link_type {
    my $self = shift;
    if ($self->rel eq 'search') {
        return "application/opensearchdescription+xml";
    }
    my $kind = $self->acquisition ? 'acquisition' : 'navigation';
    return "application/atom+xml;profile=opds-catalog;kind=$kind";
}

sub _rel_map {
    my %map = (
               new => "http://opds-spec.org/sort/new",
               popular => "http://opds-spec.org/sort/popular",
               featured => "http://opds-spec.org/featured",
               recommended => "http://opds-spec.org/recommended",
               shelf => "http://opds-spec.org/shelf",
               subscriptions => "http://opds-spec.org/subscriptions",
               crawlable => "http://opds-spec.org/crawlable",
              );
}

sub relationship {
    my $self = shift;
    my %mapped = $self->_rel_map;
    my $rel = $self->rel;
    return $mapped{$rel} || $rel;
}

sub as_link {
    my $self = shift;
    my $link = XML::Atom::Link->new(Version => 1.0);
    $link->rel($self->relationship);
    $link->href($self->prefix . $self->href);
    $link->type($self->link_type);
    # unclear
    # If no appropriate relation is found, the Feeds SHOULD use a descriptive
    # "atom:title" element and the "atom:link"s SHOULD use a descriptive "title"
    # attribute.
    if ($self->rel eq 'subsection') {
        if (my $title = $self->title) {
            $link->title($title);
        }
    }
    return $link;
}

sub identifier {
    my $self = shift;
    return $self->id || $self->prefix . $self->href;
}

sub as_entry {
    my $self = shift;
    my $item = XML::Atom::Entry->new(Version => 1.0);
    $item->title($self->title);
    $item->id($self->identifier);
    $item->content($self->description);
    $item->updated($self->_dt_formatter->format_datetime($self->updated));
    $item->add_link($self->as_link);
    return $item;
}

1;


lib/XML/OPDS/Acquisition.pm
package XML::OPDS::Acquisition;

use strict;
use warnings FATAL => 'all';
use Types::Standard qw/Str ArrayRef InstanceOf Object/;
use Moo;
use DateTime;
use XML::Atom;
use XML::Atom::Person;
use XML::Atom::Entry;

=head1 NAME

XML::OPDS::Acquisition - Acquisition elements for OPDS feeds

=head1 SETTERS/ACCESSORS

The following accessors are read-only and are meant to be passed as an
hash to the C<new> constructor.

=head2 id

Optional. The example specification uses uuid like
C<urn:uuid:2853dacf-ed79-42f5-8e8a-a7bb3d1ae6a2>. If not provided, the
C<href> (with the prefix if provided) will be used instead.

=head2 prefix

If provided, every uri will have this string prepended, so you can
just pass URIs like '/path/to/file' and have them consistently turned
to 'http://myserver.org/path/to/file' if you set this to
'http://myserver.org'. See also L<XML::OPDS> C<prefix> method.

=cut

has id => (is => 'ro', isa => Str);

has prefix => (is => 'ro', isa => Str, default => sub { '' });

=head2 href

The URI of the resource. Required.

=cut

has href => (is => 'ro', isa => Str, required => 1);

=head2 title

The title. Required.

=cut

has title => (is => 'ro', isa => Str, required => 1);

=head2 files

An arrayref with the download files. The prefix is added if set.

=head1 OPTIONAL ATTRIBUTES

The following attributes are optional and describe the publication.

=head2 authors

An arrayref of either scalars with names, or hashrefs with C<name> and
C<uri> as keys. C<uri> is optional.

=head2 summary

Plain text.

=head2 description

HTML allowed.

=head2 language

=head2 issued

The publication date.

=head2 publisher

=head2 updated

A DateTime object with the time of the last update of the resource.

=cut

has authors => (is => 'ro', isa => ArrayRef);

has summary => (is => 'ro', isa => Str);

has description => (is => 'ro', isa => Str);

has language => (is => 'ro', isa => Str);

has issued => (is => 'ro', isa => Str);

has publisher => (is => 'ro', isa => Str);

has updated => (is => 'rw', isa => InstanceOf['DateTime'],
                default => sub { return DateTime->now });

has files => (is => 'ro', isa => ArrayRef[Str], default => sub { [] });

=head2 thumbnail

The uri of the thumbnail

=head2 image

The uri of the image

=head1 METHODS

Usually they are for internal usage.

=head2 identifier

=head2 authors_as_links

Return a list of L<XML::Atom::Person> objects from the C<authors>
value.

=head2 files_as_links

Return a list of L<XML::Atom::Link> objects constructed from C<files>,
C<image>, C<thumbnail>, with the appropriate C<rel> and C<type>.

=head2 as_entry

Return the acquisition L<XML::Atom::Entry> object.

=cut

has thumbnail => (is => 'ro', isa => Str);

has image => (is => 'ro', isa => Str);

has _dc => (is => 'lazy',
            isa => Object,
            default => sub {
                XML::Atom::Namespace->new(dc => 'http://purl.org/dc/elements/1.1/');
            });

sub identifier {
    my $self = shift;
    return $self->id || $self->prefix . $self->href;
}

sub authors_as_links {
    my $self = shift;
    my @out;
    if (my $authors = $self->authors) {
        foreach my $author (@$authors) {
            my $hash = ref($author) ? $author : { name => $author };
            if (my $name = $hash->{name}) {
                my $author_obj = XML::Atom::Person->new(Version => 1.0);
                $author_obj->name($hash->{name});
                if (my $uri = $hash->{uri}) {
                    $author_obj->uri($self->prefix . $uri);
                }
                push @out,  $author_obj;
            }
        }
    }
    return @out;
}

sub files_as_links {
    my $self = shift;
    my @out;
    my %mime = (
                tex => 'application/x-tex',
                pdf => 'application/pdf',
                html => 'text/html',
                epub => 'application/epub+zip',
                muse => 'text/plain',
                txt => 'text/plain',
                zip => 'application/zip',
                png => 'image/png',
                jpg => 'image/jpeg',
                jpeg => 'image/jpeg',
                gif => 'image/gif',
                mobi => 'application/x-mobipocket-ebook',
               );
    # maybe support open-access, borrow, buy, sample, subscribe ? 8.4.1
    my @all = map { +{ rel => 'acquisition', href => $_ } } @{$self->files};
    die "Missing acquisition links" unless @all;

    if (my $thumb = $self->thumbnail) {
        push @all, { rel => 'image/thumbnail', href => $thumb };
    }
    if (my $image = $self->image) {
        push @all, { rel => 'image', href => $image };
    }
    foreach my $file (@all) {
        my $mime_type;
        if ($file->{href} =~ m/\.(\w+)\z/) {
            my $ext = $1;
            $mime_type = $mime{$ext};
        }
        next unless $mime_type;
        my $link = XML::Atom::Link->new(Version => 1.0);
        $link->rel("http://opds-spec.org/$file->{rel}");
        $link->href($self->prefix . $file->{href});
        $link->type($mime_type);
        push @out, $link;
    }
    if (@out) {
        return @out;
    }
    else {
        die "Links are required"
    };
}

sub as_entry {
    my $self = shift;
    my $entry = XML::Atom::Entry->new(Version => 1.0);
    $entry->id($self->identifier);
    $entry->title($self->title);
    $entry->updated($self->updated);
    if (my $lang = $self->language) {
        $entry->set($self->_dc, language => $lang);
    }
    foreach my $author ($self->authors_as_links) {
        $entry->add_author($author);
    }
    if (my $summary = $self->summary) {
        $entry->summary($summary);
    }
    if (my $desc = $self->description) {
        $entry->content($desc);
    }
    foreach my $link ($self->files_as_links) {
        $entry->add_link($link);
    }
    return $entry;
}

1;


lib/XML/OPDS/OpenSearch/Query.pm
package XML::OPDS::OpenSearch::Query;
use strict;
use warnings FATAL => 'all';
use Types::Standard qw/Maybe Str Enum Int/;
use Moo;
use URI::Escape ();

=head1 NAME

XML::OPDS::OpenSearch::Query - OpenSearch query element for XML::OPDS

=head1 DESCRIPTION

This module is mostly a helper to provide validation to create
C<Query> elements in an OPDS search result. Notably, this doesn't
expand to XML. The serialization is left to L<XML::OPDS>.

=head1 SYNOPSIS

  use XML::OPDS;
  use XML::OPDS::OpenSearch::Query;
  use Data::Page;
  my $feed = XML::OPDS->new;
  my $query = XML::OPDS::OpenSearch::Query->new(role => "example",
                                                searchTerms => "my terms",
                                                );
  my $pager = Data::Page->new;
  $pager->total_entries(50);
  $pager->entries_per_page(20);
  $pager->current_page(1);
  $feed->search_result_pager($pager);
  $feed->search_result_terms('my terms');
  
  # please note that if you set the search_result_pager and
  # search_result_terms a role:request query is added automatically, so
  # here we add only the example.
  
  $feed->search_result_queries([$query]);
  print $feed->render;

=head1 SETTERS/ACCESSORS

All of them are optionals except C<role>.

=head2 role

Role values (from L<http://www.opensearch.org/Specifications/OpenSearch/1.1#Role_values>)

=over 4

=item request

Represents the search query that can be performed to retrieve the same
set of search results.

=item example

Represents a search query that can be performed to demonstrate the search engine. 

=item related

Represents a search query that can be performed to retrieve similar
but different search results.

=item correction

Represents a search query that can be performed to improve the result
set, such as with a spelling correction.

=item subset

Represents a search query that will narrow the current set of search results. 

=item superset

Represents a search query that will broaden the current set of search results.

=back

=cut

has role => (is => 'rw',
             isa => Enum[qw/role request example related correction subset superset/]);

=head2 title

Contains a human-readable plain text string describing the search request.

Restrictions: The value must contain 256 or fewer characters of plain
text. The value must not contain HTML or other markup.

This object stores an arbitrary string, but cut it at 256 when
producing the attributes.

=head2 totalResults

Integer.

Contains the expected number of results to be found if the search request were made. 

=head2 searchTerms

String.

Contains the value representing the "searchTerms" as an OpenSearch 1.1 parameter.

The URI escaping is performed by the module.

=head2 count

Integer.

Replaced with the number of search results per page desired by the search client. 

=head2 startIndex

Integer.

Replaced with the index of the first search result desired by the search client. 

=head2 startPage

Integer.

Replaced with the page number of the set of search results desired by the search client. 

=head2 language

String. The value must conform to the XML 1.0 Language Identification,
as specified by RFC 5646. In addition, a value of "*" will signify
that the search client desires search results in any language.

This module passes it verbatim.

=head2 inputEncoding

The value must conform to the XML 1.0 Character Encodings, as
specified by the IANA Character Set Assignments.

This module passes it verbatim.

=head2 outputEncoding

Same as above.

=head1 METHODS

=head2 attributes_hashref

Return the attributes which are defined in an hashref. The C<title> is
mangled to 256 characters and 

=head1 SEE ALSO

Specification: L<http://www.opensearch.org/Specifications/OpenSearch/1.1>

=cut

has title => (is => 'rw', isa => Maybe[Str]);
has totalResults => (is => 'rw', isa => Maybe[Int]);
has searchTerms => (is => 'rw', isa => Maybe[Str]);
has count => (is => 'rw', isa => Maybe[Int]);
has startIndex => (is => 'rw', isa => Maybe[Int]);
has startPage => (is => 'rw', isa => Maybe[Int]);
has language => (is => 'rw', isa => Maybe[Str]);
has inputEncoding => (is => 'rw', isa => Maybe[Str]);
has outputEncoding => (is => 'rw', isa => Maybe[Str]);

sub attributes_hashref {
    my $self = shift;
    my %out = (role => $self->role);
    if (defined $self->title) {
        $out{title} = substr $self->title, 0, 256;
    }
    if (defined $self->searchTerms) {
        $out{searchTerms} = URI::Escape::uri_escape($self->searchTerms);
    }
    foreach my $accessor (qw/totalResults
                             count
                             startIndex
                             startPage
                             language
                             inputEncoding
                             outputEncoding/) {
        my $v = $self->$accessor;
        if (defined $v) {
            $out{$accessor} = $v;
        }
    }
    return \%out;
}

1;


lib/XML/OPDS.pm
package XML::OPDS;

use strict;
use warnings FATAL => 'all';
use Types::Standard qw/Str Object ArrayRef InstanceOf Maybe Int/;
use Moo;
use DateTime;
use DateTime::Format::RFC3339;
use XML::Atom;
use XML::Atom::Feed;
use XML::Atom::Entry;
use XML::OPDS::Navigation;
use XML::OPDS::Acquisition;
use XML::OPDS::OpenSearch::Query;

=head1 NAME

XML::OPDS - OPDS (Open Publication Distribution System) feed creation

=head1 VERSION

Version 0.04

=cut

our $VERSION = '0.04';


=head1 DESCRIPTION

This module facilitates the creation of OPDS feeds.

The specifications can be found at L<http://opds-spec.org/> while the
validator is at L<http://opds-validator.appspot.com/>.

The idea is that it should be enough to pass the navigation links and
the title entries with some data, and have the feed back.

The OPDS feeds are basically Atom feeds, hence this module uses
L<XML::Atom> under the hood.

Some features are not supported yet, but patches are welcome. Also
keep in mind that the applications which are supposed to talk to your
server have a level of support which varies from decent to "keep
crashing".

This is still very much a work in progress, but it's already
generating valid and usable catalogs.

=head1 SYNOPSIS

  use XML::OPDS;
  my $feed = XML::OPDS->new(prefix => 'http://amusewiki.org');
  # add two links, self and start are mandatory.
  $feed->add_to_navigations(
                            rel => 'self',
                            title => 'Root',
                            href => '/',
                           );
  $feed->add_to_navigations(
                            rel => 'start',
                            title => 'Root',
                            href => '/',
                           );
  # add a navigation for the title list, marking as leaf, where the
  # download links can be retrieved.
  $feed->add_to_navigations(
                            title => 'Titles',
                            description => 'texts sorted by title',
                            href => '/titles',
                            acquisition => 1,
                           );
  # and render
  print $feed->render;
  # you can reuse the object for leaf feeds (i.e. the acquistion
  # feeds), pushing another self navigation, which will replace the
  # previous one.
  $feed->add_to_navigations(
                            rel => 'self',
                            title => 'Titles',
                            description => 'texts sorted by title',
                            href => '/titles',

                           );
  # or, implicitely setting the self rel and cleaning the navigation
  # stash, keeping the meta
  $feed->add_to_navigations_new_level(
                                      title => 'Titles',
                                      acquisition => 1,
                                      href => '/titles',
                                     );
  $feed->add_to_acquisitions(
                             href => '/my/title',
                             title => 'My title',
                             files => [ '/my/title.epub' ],
                            );
  # and here we have an acquisition feed, because of the presence of
  # the acquisition.
  print $feed->render;

=head1 ENCODING

Even if the module wants characters as input (decoded strings, not
bytes), the output XML is an UTF-8 encoded string.

=head1 SETTERS/ACCESSORS

=head2 navigations

Arrayref of L<XML::OPDS::Navigation> objects. An object with a rel
C<self> (the feed itself) and one with the rel C<start> (the root
feed) are mandatory. If not present, the module will crash while
rendering the feed.

=head2 acquisitions

Arrayref of L<XML::OPDS::Acquisition> objects. If one or more objects
are present, the feed will become an acquistion feed.

=head2 author

The producer of the feed. Defaults to this class name and version.

=head2 author_uri

The uri of the author. Defaults to L<http://amusewiki.org> (which is
the home of this class).

=head2 prefix

Default to the empty string. On instances of this class, by itself has
no effect. However, when calling C<add_to_acquisitions> and
C<add_to_navigations>, it will be passed to the constructors of those
objects.

This is usually the hostname of the OPDS server. So you need just to
pass, e.g. 'http://amusewiki.org' and have all the links prefixed by
that (no slash mangling or adding is performed). If you are going to
pass the full urls, leave it at the default.

=head2 updated

Default to current timestamp. When calling C<create_navigation> or
C<create_acquistion>, use this timestamp as default.

=head2 logo

The feed logo. If prefix is set, prepend it.

=head2 icon

The feed icon. If prefix is set, prepend it.

=head1 METHODS

=head2 render

Return the generated xml.

=head2 atom

Return the L<XML::Atom::Feed> object.

=head2 create_navigation(%args)

Create a L<XML::OPDS::Navigation> object inheriting the prefix.

=head2 create_acquisition(%args)

Create a L<XML::OPDS::Acquisition> object inheriting the prefix.

=head2 add_to_navigations(%args)

Call C<create_navigation> and add it to the C<navigations> stack.

=head2 add_to_navigations_new_level(%args)

Like C<add_to_navigations>, but it's meant to be used for
C<rel:self> elements.

The C<rel:self> attribute is injected in the arguments which are
passed to C<create_navigation>.

If a navigation with the attribute C<rel> set to C<self> was already
present in the stack, the new one will become the new C<self>, while
the old one will become an C<up> rel.

Also, this will remove any existing navigation with the C<rel>
attribute set to C<subsection>, given that you are creating a new
level.

This is designed to play well with chained actions (so you can reuse
the object, stack selfs, and the result will be correct).

=head2 add_to_acquisitions(%args)

Call C<create_acquisition> and add it to the C<acquisition> stack.

=head1 INTERNAL METHODS

=head2 navigation_entries

Return a list of L<XML::OPDS::Navigation> objects excluding unique
relationships like C<self>, C<start>, C<up>, C<previous>, C<next>,
C<first>, C<last>.

=head2 navigation_hash

Return an hashref, where the keys are the C<rel> attributes of the
navigation objects. The value is an object if the navigation is meant
to be unique, or an arrayref of objects if not so.

=head2 is_acquisition

Return true if there are acquisition objects stacked.

=cut

has navigations => (is => 'rw',
                    isa => ArrayRef[InstanceOf['XML::OPDS::Navigation']],
                    default => sub { [] });
has acquisitions => (is => 'rw',
                     isa => ArrayRef[InstanceOf['XML::OPDS::Acquisition']],
                     default => sub { [] });
has author => (is => 'rw', isa => Str, default => sub { __PACKAGE__ . ' ' . $VERSION });
has author_uri => (is => 'rw', isa => Str, default => sub { 'http://amusewiki.org' });
has prefix => (is => 'rw', isa => Str, default => sub { '' });
has updated => (is => 'rw', isa => Object, default => sub { DateTime->now });
has icon => (is => 'rw', isa => Str, default => sub { '' });
has logo => (is => 'rw', isa => Str, default => sub { '' });

has _dt_formatter => (is => 'ro', isa => Object,
                      default => sub { DateTime::Format::RFC3339->new });
has _fh => (is => 'ro',
            isa => Object,
            default => sub {
                XML::Atom::Namespace->new(fh => 'http://purl.org/syndication/history/1.0');
            });

# opensearch accessors

has _os => (is => 'ro',
            isa => Object,
            default => sub {
                XML::Atom::Namespace->new(opensearch => 'http://a9.com/-/spec/opensearch/1.1/');
            });


=head1 OPENSEARCH RESULTS

The following attributes can be set if you are building an Atom
response for OpenSearch. See L<XML::OPDS::OpenSearch::Query> for a
concrete example.

=head2 search_result_pager

A L<Data::Page> object with the specification of the pages.

=head2 search_result_terms

A string with the query for which you are serving the results.

=head2 search_result_queries

Additional Query elements, should be an arrayref of
L<XML::OPDS::OpenSearch::Query> objects.

=cut

has search_result_pager => (is => 'rw',
                            isa => InstanceOf['Data::Page']);

has search_result_terms => (is => 'rw',
                            isa => Str);

has search_result_queries => (is => 'rw',
                              isa => ArrayRef[InstanceOf['XML::OPDS::OpenSearch::Query']],
                              default => sub { [] },
                             );

sub navigation_entries {
    my $self = shift;
    my $hash = $self->navigation_hash;
    my @others;
    foreach my $k (sort keys %$hash) {
        my $entries = $hash->{$k};
        # exclude the uniques
        if (ref($entries) eq 'ARRAY') {
            push @others, @$entries;
        }
    }
    return @others;
}

sub navigation_hash {
    my $self = shift;
    my $navs = $self->navigations;
    die "Missing navigations" unless $navs && @$navs;
    my %out;
    my %uniques = (
                   start => 1,
                   self => 1,
                   up => 1,
                   next => 1,
                   previous => 1,
                   first => 1,
                   last => 1,
                   search => 1,
                   crawlable => 1,
                  );
    foreach my $nav (@$navs) {
        my $rel = $nav->rel;
        # uniques
        if ($uniques{$rel}) {
            $out{$rel} = $nav;
        }
        else {
            $out{$rel} ||= [];
            push @{$out{$rel}}, $nav;
        }
    }
    return \%out;
}

sub is_acquisition {
    if (my $acquisitions = shift->acquisitions) {
        return scalar(@$acquisitions);
    }
    else {
        return 0;
    }
}

sub _is_paged {
    my $self = shift;
    my $partial = 0;
    foreach my $nav (@{$self->navigations}) {
        if ($nav->rel =~ m/\A(next|previous|first|last)\z/) {
            $partial = 1;
            last;
        }
    }
    return $partial;
}

sub atom {
    my $self = shift;
    my $feed = XML::Atom::Feed->new(Version => 1.0);
    my $navs = $self->navigation_hash;
    my $main = delete $navs->{self};
    die "Missing self navigation element!" unless $main;
    $feed->id($main->identifier);
    $feed->add_link($main->as_link);
    my @nav_entries;
    foreach my $rel (sort keys %$navs) {
        # use only the unique
        my $nav = delete $navs->{$rel};
        if (ref($nav) eq 'ARRAY') {
            push @nav_entries, @$nav;
        }
        else {
            $feed->add_link($nav->as_link);
        }
    }
    $feed->title($main->title);
    $feed->updated($self->_dt_formatter->format_datetime($main->updated));
    if (my $icon = $self->icon) {
        $feed->icon($self->prefix . $icon);
    }
    if (my $logo = $self->logo) {
        $feed->logo($self->prefix . $logo);
    }
    if (my $author_name = $self->author) {
        my $author = XML::Atom::Person->new(Version => 1.0);
        $author->name($author_name);
        if (my $author_uri = $self->author_uri) {
            $author->uri($author_uri);
        }
        $feed->author($author);
    }

    # opensearch element
    # http://www.opensearch.org/Specifications/OpenSearch/1.1#OpenSearch_response_elements
    if (my $pager = $self->search_result_pager) {
        $feed->set($self->_os, totalResults =>  $pager->total_entries);
        $feed->set($self->_os, startIndex => $pager->first);
        $feed->set($self->_os, itemsPerPage => $pager->entries_per_page);
        if (my $term = $self->search_result_terms) {
            my $query = XML::OPDS::OpenSearch::Query->new(
                                                          role => 'request',
                                                          searchTerms => $term,
                                                         );
            $feed->add($self->_os, Query => undef, $query->attributes_hashref);
        }
    }
    foreach my $query (@{ $self->search_result_queries }) {
        $feed->add($self->_os, Query => undef, $query->attributes_hashref);
    }
    if ($self->is_acquisition) {
        # if it's an acquisition feed, stuff the links in the feed,
        # but filter out the subsections. And probably other stuff as well.
        foreach my $link (@nav_entries) {
            my %rels = (related => 1, alternate => 1);
            $feed->add_link($link->as_link) if $rels{$link->rel};
        }
        unless ($self->_is_paged) {
            $feed->set($self->_fh, complete => undef);
        }
        foreach my $entry (@{$self->acquisitions}) {
            $feed->add_entry($entry->as_entry);
        }
    }
    else {
        # othewise use the links to create entries
        foreach my $entry (@nav_entries) {
            $feed->add_entry($entry->as_entry);
        }
    }
    return $feed;
}

sub render {
    shift->atom->as_xml;
}

sub create_navigation {
    my $self = shift;
    return XML::OPDS::Navigation->new(prefix => $self->prefix,
                                      updated => $self->updated,
                                      @_);
}

sub add_to_navigations {
    my $self = shift;
    my $navigation = $self->create_navigation(@_);
    push @{$self->navigations}, $navigation;
    return $navigation;
}

sub add_to_navigations_new_level {
    my $self = shift;
    my $navigation = $self->create_navigation(rel => 'self', @_);
    # turn the previous self in an "up" link.
    if ($navigation->rel eq 'self') {
        # new level, so remove the subsections
        my @existing = grep { $_->rel ne 'subsection' } @{$self->navigations};
        # promote the existing self to "up"
        foreach my $previous (grep { $_->rel eq 'self' } @existing) {
            $previous->rel('up');
        }
        # reset
        $self->navigations(\@existing);
    }
    push @{$self->navigations}, $navigation;
    return $navigation;
}

sub create_acquisition {
    my $self = shift;
    return XML::OPDS::Acquisition->new(prefix => $self->prefix,
                                       updated => $self->updated,
                                       @_);
}

sub add_to_acquisitions {
    my $self = shift;
    my $acquisition = $self->create_acquisition(@_);
    push @{$self->acquisitions}, $acquisition;
    return $acquisition;
}


=head1 AUTHOR

Marco Pessotto, C<< <melmothx at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-xml-opds at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=XML-OPDS>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc XML::OPDS


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=XML-OPDS>

=item * MetaCPAN

L<http://metacpan.org/pod/XML::OPDS>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2016 Marco Pessotto.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.


=cut

1; # End of XML::OPDS
